### ISSUE TITLE
Speed testing app can't measure upload speed

### ISSUE THREAD

**Issue by vozersky** [AdguardTeam/AdguardForAndroid#1249]
Happens **only** with HTTP proxy filtering mode (Automatic) on rooted device
All other settings are default.

[This ](https://play.google.com/store/apps/details?id=com.staircase3.opensignal&hl=en) is the app.

Debug log: [adguard.log.zip](https://github.com/AdguardTeam/AdguardForAndroid/files/1039016/adguard.log.zip)

STR: 
* hit the "Test" button
* 0 kb/s upload speed will be displayed after the test

<details>

![image](https://cloud.githubusercontent.com/assets/17472907/26594990/85214a50-4572-11e7-812e-70064739bbc0.png)

</details>
<br>

**Comment by nkartyshov** [comment]
Problem reproduced without AG. 
@vozersky Can you check app again, please?

**Comment by ameshkov** [comment]
What should we do with this?

**Comment by nkartyshov** [comment]
The application sends an invalid http request to check the download speed, because of  this it is shown equal to 0 kb. 

Unfortunately, fixing the Http request did not help, as did the application exceptions from filtering.

**Comment by ameshkov** [comment]
Solution: disable AG for this app in the apps management section

**Comment by carlylemiii** [comment]
@ameshkov Do you have any comments I can forward to OpenSignal in hopes they can fix it on their end? This is the one thing keeping me from using proxy mode, since a reboot is the only fix (i.e., not disabling AdGuard).

**Comment by ameshkov** [comment]
@carlylemiii 

> Do you have any comments I can forward to OpenSignal in hopes they can fix it on their end? 

I doubt they will bother :( The problem is in the redirect itself, not in the way we handle HTTP (the request might be not valid, but it does not matter -- we tried bypassing data).

Also, disabling AG for the app will also help.

**Comment by carlylemiii** [comment]
@ameshkov 

> Also, disabling AG for the app will also help.

You mean that for you, with auto proxy mode, if AdGuard filtering is disabled for OpenSignal, the upload portion of the speed test works? I totally don't care about ad-blocking for that app, so I'm cool with disabling routing altogether, but for me, that doesn't do it - the upload speed is still 0. I am on the 2.10.56 beta.

**Comment by ameshkov** [comment]
Huh, stupid us, we should've not analyzed the app traffic, but everything else.

The problem seems to be in dropping QUIC traffic system-wide:
```
iptables -A OUTPUT -p udp --dport 443 -j DROP";
```

@nkartyshov can't we use `-t nat` for it? In this case adding an exception will also unblock QUIC for that app.

**Comment by Revertron** [comment]
This command works perfectly:
`iptables -A OUTPUT -p udp --dport 443 -m owner --uid-owner <UID> -j DROP`

**Comment by ameshkov** [comment]
@Revertron it sure does. The point is that we don't need to narrow QUIC ban to a particular app. On the contrary, we want to block it system wide, **save for the apps AG is disabled for**.

**Comment by Revertron** [comment]
Two commands then:
```
iptables -A OUTPUT -p udp --dport 443 -m owner --uid-owner <uid> -j RETURN
iptables -A OUTPUT -p udp --dport 443 -j DROP
```
First one enables traffic for one user, second - disables it for others.

**Comment by ameshkov** [comment]
@Revertron if we use `-t nat` we won't need more commands as exceptions are added to the nat table

**Comment by Revertron** [comment]
We can not filter with -t nat:

iptables v1.6.0:
The "nat" table is not intended for filtering, the use of DROP is therefore inhibited.

**Comment by sfionov** [comment]
quic may also be matched by rule like -m u32 --u32 "0x9&0xffff3030=0x51303030" --dport 443
but older kernels doesn't have u32

**Comment by sfionov** [comment]
more accurate rule is --u32 "0x0&0xfb=0x9 && 0x9&0xffff3030=0x51303030"

**Comment by ameshkov** [comment]
> The "nat" table is not intended for filtering, the use of DROP is therefore inhibited.

You can use `REDIRECT` instead (and route it to some non-existing place)..

@sfionov that's interesting, and what will happen on older kernels where these are not supported?

**Comment by sfionov** [comment]
Rule just will not be added. However, we may use `||` in shell commands for fallback of particular command.
`-m u32` is used by `netd` since Android 6.0.

**Comment by Revertron** [comment]
I propose blocking all UDP to ports 80 & 443 (as we do now) on older platforms and QUIC (with -m u32) on new ones.

**Comment by ameshkov** [comment]
Let's use the `||` approach then

**Comment by nkartyshov** [comment]
fixed

**Comment by carlylemiii** [comment]
I don't know if this is related or should be its own issue, but with proxy mode I'm getting a ridiculously low ping, one that isn't even typical of LTE. It's weird. The normal double digits pings are without proxy mode filtering, and the single digits are with proxy filtering.

https://www.dropbox.com/sh/l5lpuoctpm90sb1/AAAhJz3bp2YKZPPlxOSmbj_ya?dl=0

That said, it's so awesome to see upload working now!!

**Comment by ameshkov** [comment]
> The normal double digits pings are without proxy mode filtering, and the single digits are with proxy filtering.

That's the downside of how filtering works, connection with the local proxy occurs immediately
